addons/udes_stock/models/product_template.py

```diff
@@ -49,3 +49,24 @@ def _domain_height_category(self):
     def unlink(self):
         """Override superclass to prevent deletion."""
         raise ValidationError(_("Products may not be deleted. Please archive them instead."))
+
+    @api.onchange("tracking")
+    def onchange_tracking(self):
```

Comment:
this and `constrain_tracking()` do the same thing, we can abstract the contents to a helper function

Reply:
```python
    @api.onchange("tracking")
    def onchange_tracking(self):
        self.check_tracking()

    @api.constrains("tracking")
    def constrain_tracking(self):
        self.check_tracking()

    def check_tracking(self):
        for product in self:
            if product.product_variant_ids.has_goods_in_transit_or_stock():
                # If there is stock, raise an error to prevent changing the tracking
                raise ValidationError(
                    _(
                        "Cannot change tracking for product '%s' with stock or move lines in ready state."
                    )
                    % product.name
                )
```